[JSC] Untie emitAwait() from emitYield() to reduce bytecode size of async functions

shvaikalesh opened the pull request:
#### 402711f086a1390c86b2b4dcd21989b0f6a305fb
<pre>
[JSC] Untie emitAwait() from emitYield() to reduce bytecode size of async functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=263031">https://bugs.webkit.org/show_bug.cgi?id=263031</a>
&lt;rdar://problem/116825631&gt;

Reviewed by Yusuke Suzuki.

Since async functions are implemented through generators for convenience, `await` keyword shares
the implementation with `yield`.

However, unlike `yield` in generators, `await` can&apos;t produce a `RETURN` completion [1], so this
patch reduces its bytecode size by simplifying ResumeMode checks. The similar change is made
to the prelude of async functions as well.

Another reason to separate implementations is that, for spec compliance [2], `yield*` in a sync
generator must reuse the received iterator result object, thus the regular `yield` would need to
wrap the value [3] in an iterator result object as well, which is unsensible to do for async
generators / functions and would bloat their bytecode size.

Also, this change a) removes unconditional (even in case of ignored result) move() from emitAwait()
and b) hoists `move(value.get(), generatorValueRegister())` in `yield*` implementation to colocate
with ResumeMode checks, making the code easier to follow and emit one less move().

No new tests, no behavior change.
This patch offers 5.1% bytecode reduction on JetStream3/async-fs subtest.

[1]: <a href="https://tc39.es/ecma262/#await">https://tc39.es/ecma262/#await</a>
[2]: <a href="https://bugs.webkit.org/show_bug.cgi?id=174756">https://bugs.webkit.org/show_bug.cgi?id=174756</a>
[3]: <a href="https://tc39.es/ecma262/#sec-createiterresultobject">https://tc39.es/ecma262/#sec-createiterresultobject</a>

* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::emitYield):
(JSC::BytecodeGenerator::emitAwait):
(JSC::BytecodeGenerator::emitDelegateYield):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
(JSC::BytecodeGenerator::emitAwait):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::FunctionNode::emitBytecode):
(JSC::YieldExprNode::emitBytecode):
(JSC::AwaitExprNode::emitBytecode):

Canonical link: <a href="https://commits.webkit.org/269321@main">https://commits.webkit.org/269321@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/703611c3ea81fb116c5e822d8667aa9536f9b557

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22219 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22423 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23291 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24116 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/20571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22467 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26680 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22747 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21587 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22452 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22165 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19278 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24969 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19222 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20134 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26386 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/19322 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20233 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20363 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24249 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/21591 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20800 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/17699 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/25637 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20152 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/19943 "Passed tests") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6030 "Failed to checkout and rebase branch from PR 19012") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5291 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24363 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/26911 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20748 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5863 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
```diff
@@ -4908,6 +4908,20 @@ RegisterID* BytecodeGenerator::emitYield(RegisterID* argument, JSAsyncGenerator:
+    emitThrow(generatorValueRegister());
+
+    emitLabel(normalLabel.get());
+    return move(dst, generatorValueRegister());
```
How about checking `if (dst == ignoredResult())` and skip the move instead of calling `generator.finalDestination` in the caller side?

shvaikalesh replied:
I've removed `generator.finalDestination()` from the caller and instead added handling of `dst == nullptr`, ensuring that no extra `move` is emitted for `ignoredResult()`.

We can just
```cpp
return move(dst, generatorValueRegister());
```

here because `BytecodeGenerator::move()` checks for `ignoredResult`.

Constellation commented on Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
```diff
@@ -4908,6 +4908,20 @@ RegisterID* BytecodeGenerator::emitYield(RegisterID* argument, JSAsyncGenerator:
 
+RegisterID* BytecodeGenerator::emitAwait(RegisterID* dst, RegisterID* src)
+{
+    ASSERT(dst != ignoredResult());
```
And then, it is not necessary.

Constellation reviewed (approved):
r=me  with comment.

Constellation reviewed (approved):
r=me

Commit:
[JSC] Untie emitAwait() from emitYield() to reduce bytecode size of async functions
https://bugs.webkit.org/show_bug.cgi?id=263031
<rdar://problem/116825631>

Reviewed by Yusuke Suzuki.

Since async functions are implemented through generators for convenience, `await` keyword shares
the implementation with `yield`.

However, unlike `yield` in generators, `await` can't produce a `RETURN` completion [1], so this
patch reduces its bytecode size by simplifying ResumeMode checks. The similar change is made
to the prelude of async functions as well.

Another reason to separate implementations is that, for spec compliance [2], `yield*` in a sync
generator must reuse the received iterator result object, thus the regular `yield` would need to
wrap the value [3] in an iterator result object as well, which is unsensible to do for async
generators / functions and would bloat their bytecode size.

Also, this change a) removes unconditional (even in case of ignored result) move() from emitAwait()
and b) hoists `move(value.get(), generatorValueRegister())` in `yield*` implementation to colocate
with ResumeMode checks, making the code easier to follow and emit one less move().

No new tests, no behavior change.
This patch offers 5.1% bytecode reduction on JetStream3/async-fs subtest.

[1]: https://tc39.es/ecma262/#await
[2]: https://bugs.webkit.org/show_bug.cgi?id=174756
[3]: https://tc39.es/ecma262/#sec-createiterresultobject

* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::emitYield):
(JSC::BytecodeGenerator::emitAwait):
(JSC::BytecodeGenerator::emitDelegateYield):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
(JSC::BytecodeGenerator::emitAwait):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::FunctionNode::emitBytecode):
(JSC::YieldExprNode::emitBytecode):
(JSC::AwaitExprNode::emitBytecode):

Canonical link: https://commits.webkit.org/269321@main
